openedx/openedx-platform
# feat: ESG BFF submission/lock endpoint

## Description

Add the `submission/lock` endpoint in support of enhanced staff grader (ESG).

JIRA: [AU-366](https://openedx.atlassian.net/browse/AU-366)
Prerequisite PR: https://github.com/edx/edx-ora2/pull/1741

FYI @edx/masters-devs-gta 

## Testing instructions

- [ ] Test suites pass

After pulling prerequisite work from above PR, exercise the following endpoint:

```
POST {{protocol}}://{{lms_url}}/api/ora_staff_grader/submission/lock?submissionId={{submission_id}}&value={{true/false}}&ora_location={{block_id_encoded}}
```
- [ ] Missing query params returns a `400`
- [ ] Bad ORA location returns a `400`

**Note:** testing requires presence of valid `X-CSRFToken` header.

Operating with `value=true` param instructs the endpoint to try to **claim a submission lock**. Verify the following:
- [ ] Attempting to claim a lock for a submission without a current lock returns successfully `200 - {'lockStatus': 'in-progress'}`
- [ ] Attempting to claim a lock for a submission where you own the current lock (within the past 8 hours) returns successfully `200 - {'lockStatus': 'in-progress'}`
- [ ] Attempting to claim a lock for a submission recently locked by another user returns an error `403 - {'error_code': 'ERR_LOCK_CONTESTED'}`

Operating with `value=false` param instructs the endpoint to try to **delete a submission lock**. Verify the following:
- [ ] Attempting to delete a lock for a submission without a current lock (which is a no-op but) returns successfully `200 - {'lockStatus': 'unlocked'}`
- [ ] Attempting to delete a lock for a submission where you own the current lock (within the past 8 hours) returns successfully `200 - {'lockStatus': 'unlocked'}`
- [ ] Attempting to delete a lock for a submission recently locked by another user returns an error `403 - {'error_code': 'ERR_LOCK_CONTESTED'}`
